common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+     *
+     */
+    private def makeNoteOfCacheHit(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_HIT, s"[GET] serving from cache: $key")(logger)
+    }
+
+    /**
+     * Log a cache miss
+     *
+     */
+    private def makeNoteOfCacheMiss(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_MISS, s"[GET] serving from datastore: $key")(logger)
+    }
+
+    /**
+     * We have initiated a read (in cacheLookup), now handle its completion
+     *
+     */
+    private def listenForReadDone(key: Any, entry: Entry, future: => Future[W], promise: Promise[W])(
+        implicit logger: Logging, ec: ExecutionContext): Unit = {
+
+        future onComplete {
+            case Success(value) => {
+                //
+                // if the datastore read was successful, then transition to the Cached state
+                //
+                logger.debug(this, "Read backend part done, now marking cache entry as done")
+                if (entry.readDone()) {
+                    promise success value
+                } else {
+                    // if this ever happens, this cache impl is buggy
+                    readOops(key, entry, promise, ConcurrentOperationUnderRead(entry.state.get))
+                }
+            }
+            case Failure(t) => {
+                //
+                // oops, the datastore read failed. invalidate the cache entry
+                //
+                // note that this might be a perfectly legitimate failure,
+                // e.g. a lookup for a non-existant key; we need to pass the particular t through
+                //
+                readOops(key, entry, promise, t)
+            }
+        }
+    }
+
+    /**
+     * We have initiated a write, now handle its completion
+     *
+     */
+    private def listenForWriteDone(key: Any, entry: Entry, future: Future[Winfo])(
+        implicit logger: Logging, ec: ExecutionContext): Unit = {
+
+        future onComplete {
+            case Success(docinfo) => {
+                //
+                // if the datastore write was successful, then transition to the Cached state
+                //
+                logger.debug(this, "Write backend part done, now marking cache entry as done")
+
+                if (!entry.writeDone()) {
+                    // if this ever happens, this cache impl is buggy
+                    logger.error(this, ConcurrentOperationUnderUpdate.toString())
+                    writeOops(key, entry)
+
+                } else {
+                    logger.debug(this, s"Write all done $key ${entry.transid} ${entry.state.get}")
+                }
+            }
+            case Failure(t) => {
+                //
+                // oops, the datastore write failed. invalidate the cache entry
+                //
+                writeOops(key, entry)
+            }
+        }
```

Comment:
Same as above, this can be simplified to reduce duplicated code:

```
future.flatMap { value =>
    logger.debug(this, "Write backend part done, now marking cache entry as done")
    if (entry.writeDone()) {
        logger.debug(this, s"Write all done $key ${entry.transid} ${entry.state.get}")
        Future.successful(value)
    } else {
        Future.failed(ConcurrentOperationUnderUpdate(entry.state.get))
    }
} onFailure {
    case _ =>
        entry.invalidate
        cache remove key
}
```

WriteOops can be inlined, and the whole method could be inlined aswell, keeping the relevant code local to `cacheUpdate`. Thoughts?